src/modsim2/similarity/metrics/kde.py

```diff
@@ -0,0 +1,295 @@
+import logging
+
+import numpy as np
+from scipy import integrate
+from sklearn.neighbors import KernelDensity, NearestNeighbors
+from torch import randint
+
+from modsim2.similarity.embeddings import EMBEDDING_FN_DICT
+
+from .metrics import DistanceMetric
+
+# Set module logger
+logger = logging.getLogger(__name__)
+
+
+class KDE(DistanceMetric):
+    """
+    Class to calculate the distance between two datasets
+    using kernel density estimation (KDE). The density estimation
+    is calculated for the source and target (A and B) datasets.
+    A metric is then applied to calculate a distance between the
+    density distributions.
+
+    sklearn's KernelDensity class has been used as a variety of
+    kernels are implemented and the bandwidth can be estimated
+    using Scott's Rule.
+    """
+
+    def __init__(self, seed: int) -> None:
+        super().__init__(seed)
+
+    def _pre_process_data(self, data_A: np.ndarray, data_B: np.ndarray):
+        pass
+
+    @staticmethod
+    def l2(
```

Comment:
`l2`, `kl`, `tv` looks like they can be refactored to share code and each provide their own inner `func`?

Reply:
Done